torch/autograd/functional.py

```diff
@@ -0,0 +1,717 @@
+import torch
+
+# Utility functions
+
+def _as_tuple(inp, arg_name, fn_name):
+    # Ensures that inp is a tuple of Tensors
+    # Returns whether or not the original inp was a tuple and the tupled version of the input
+    is_inp_tuple = True
+    if not isinstance(inp, tuple):
+        inp = (inp,)
+        is_inp_tuple = False
+
+    for i, el in enumerate(inp):
+        if not torch.is_tensor(el):
+            raise TypeError("The {} given to {} must be either a Tensor or a tuple of Tensors but the"
+                            " value at index {} has type {}.".format(arg_name, fn_name, i, type(el)))
+
+    return is_inp_tuple, inp
+
+def _tuple_postprocess(res, to_unpack):
+    # Unpacks a potentially nested tuple of Tensors
+    # to_unpack should be a single boolean or a tuple of two booleans.
+    # It is used to:
+    # - invert _as_tuple when res should match the inp given to _as_tuple
+    # - optionally remove nesting of two tuples created by multiple calls to _as_tuple
+    if isinstance(to_unpack, tuple):
+        assert len(to_unpack) == 2
+        if not to_unpack[1]:
+            res = tuple(el[0] for el in res)
+        if not to_unpack[0]:
+            res = res[0]
+    else:
+        if not to_unpack:
+            res = res[0]
+    return res
+
+def _grad_preprocess(inputs, create_graph, need_graph):
+    # Preprocess the inputs to make sure they require gradient
+    # inputs is a tuple of Tensor to preprocess
+    # create_graph specifies if the user wants gradients to flow back to the Tensors in inputs
+    # need_graph specifies if we internally want gradients to flow back to the Tensors in res
+    # Note that we *always* create a new Tensor object to be able to see the difference between
+    # inputs given as arguments and the same Tensors automatically captured by the user function.
+    # Check this issue for more details on how that can happen: https://github.com/pytorch/pytorch/issues/32576
+    res = []
+    for inp in inputs:
+        if create_graph and inp.requires_grad:
+            # Use .view_as() to get a new Tensor in a differentiable way
+            res.append(inp.view_as(inp))
```

Comment:
won't this not work if `inp` is a sparse tensor?

Reply:
Good point.
Do we have any operation that would return a view of the same sparse tensor? Or do we have to use `.clone()`?

Reply:
I don't believe we do, unfortunately.  So I think you'd have to use clone for now.